[mono][interp] Add instrinsics for common Vector128 operations

BrzVlad opened the pull request:
We add intrinsics for Vector128 intrinsics that are actively used within our bcl.

We declare a set of simd method names, the same way we do it with jit, in `simd-methods.def`. In `transform-simd.c` we lookup method names in the list of supported intrinsics for `Vector128` and `Vector128<T>`. Once we find a supported instrinsic, we generate code for it, typically a `MINT_SIMD_INTRINS_*` opcode. In order to avoid adding too many new opcodes to the interpreter, simd intrinsics are grouped by signature. So all simd intrinsics that receive a single argument and return a value, will be called through `MINT_SIMD_INTRINS_P_P`. This instruction will receive an index to get the intrinsic implementation and calls it indirectly.

Some of the intrinsics are implemented using the standard vector intrinsics, supported by gcc and clang. These do not fully expose the SIMD capabilities, so some intrinsics are implemented naively. This should still be faster than using nonvectorized approach from managed code. In the future we can add better implementation, on platforms where we have low level support. This would both be faster and reduce code size.

Vectorized algorithms that I've tested on interpreter tend to be about 1.5x - 2x faster than their non-vectorized version. On wasm, the jiterpreter handles pretty well non-vectorized algorithms (since they are typically a set of simple instructions that all get compiled to a single trace) and this PR doesn't produce a clear perf improvement. Wasm should get an additional speedup once SIMD support is added into the jiterpreter.

BrzVlad committed:
[mono][interp] Add intrinsics for most common V128 operations

We add intrinsics for Vector128 intrinsics that are actively used within our bcl.

We declare a set of simd method names, the same way we do it with jit, in `simd-methods.def`. In `transform-simd.c` we lookup method names in the list of supported intrinsics for `Vector128` and `Vector128<T>`. Once we find a supported instrinsic, we generate code for it, typically a `MINT_SIMD_INTRINS_*` opcode. In order to avoid adding too many new opcodes to the interpreter, simd intrinsics are grouped by signature. So all simd intrinsics that receive a single argument and return a value, will be called through `MINT_SIMD_INTRINS_P_P`. This instruction will receive an index to get the intrinsic implementation and calls it indirectly.

Some of the intrinsics are implemented using the standard vector intrinsics, supported by gcc and clang. These do not fully expose the SIMD capabilities, so some intrinsics are implemented naively. This should still be faster than using nonvectorized approach from managed code. In the future we can add better implmentation, on platforms where we have low level support. This would both be faster and reduce code size.

BrzVlad committed:
[mono][interp] Move defines from transform.c

BrzVlad committed:
[mono][interp] Move more defines from transform.c

BrzVlad committed:
[mono][interp] Add option to disable simd intrinsics

BrzVlad committed:
[mono][interp] Disable simd intrinsics by default on wasm

These intrinsics are not yet implemented on jiterpreter, making it slighty slower instead.

BrzVlad committed:
[mono][interp] Replace v128_create with v128_ldc if possible

v128_create receives as an argument every single element of the vector. This method is typically used with constants. For a Vector128<short> this means that creating a constant vector required 8 ldc.i4 and a v128_create. We can instead use a single instruction and embed the vector value in the code stream directly.

tannergooding commented on src/mono/mono/mini/interp/interp-simd.c:
```diff
@@ -0,0 +1,605 @@
+
+// op_Equality
+static void
+interp_v128_op_equality (gpointer res, gpointer v1, gpointer v2)
```
Should this qualify that it's bitwise equality?

It will be incorrect for floating-point if that's added in the future, for example

tannergooding commented on src/mono/mono/mini/interp/interp-simd.c:
```diff
@@ -0,0 +1,605 @@
+
+// op_Division
+static void
+interp_v128_i1_op_division (gpointer res, gpointer v1, gpointer v2)
```
No platform actually provides simd integer division today, so these ones may be unnecessary and not get the right level of testing.

tannergooding commented on src/mono/mono/mini/interp/interp-simd.c:
```diff
@@ -0,0 +1,605 @@
+static void
+interp_v128_i1_op_left_shift (gpointer res, gpointer v1, gpointer s1)
+{
+	*(v128_i1*)res = *(v128_i1*)v1 << *(gint32*)s1;
```
C's behavior around shifting doesn't necessarily match that of C#'s or the underlying platform's.

Might be worth validating this has the expected behavior for each `T`.

BrzVlad replied:
Are you referring to shifting of negative numbers ? In mono interpreter we actually use the C shifting operator and haven't encountered any problem on any platform so far. I also tested locally this vectorized code for all integer types and I found no issues.

tannergooding replied:
C# and by extension most shifting APIs exposed by .NET explicitly mask the input bits. That is, C# strictly defines shifting for `int` as `x << (y & 31)`. It likewise defines it for `long` as `x << (y & 63)`. For small types, it implements it in terms of `int` and then downcasts since it doesn't typically have "proper" support for small types.

Thus `x << 33` is treated as `x << 1` and does not return `0`. Likewise, `x << -1` is treated as `x << 31`, it is not treated as `x >> 1`.

x86/x64 define the shifting behavior as masking as well, so this tends to "just work" for `int`/`long`. But, in C/C++ it is strictly "undefined behavior" and so MSVC/GCC/Clang compilers do not have to match the underlying platform or be consistent with eachother, can make differing optimizations when constant folding or based on range propagation, etc.

Being explicit in these operations helps ensure there isn't any weirdness or inconsistency with the overshifting or negative shifting case.

Additionally, since `Vector###<T>` and generic math have "proper" support for small types, we don't implement them in terms of int and thus `Vector128<byte>.ShiftLeft` is defined as `x << (y & 7)` per element.

BrzVlad replied:
I'm quite confused. From what I understand from you, the shifting in c# might by different from the shifting done by the vectorized assembly operations behind the scenes. So I wrote a testcase to verify shifting for every single possible sbyte value. The problem is that, for negative and overflowing shift amounts, this test fails on all configurations, including on mono interpreter without this PR as well as on CoreCLR (I tested 7.0 rc1). I hope I didn't do anything stupid in this testcase, but for normal shifts between 0 and 7, the test passes just fine.

```
using System;
using System.Runtime.Intrinsics;

public class Program {

	public static bool CheckValue (Vector128<sbyte> vec, sbyte[] values, int index, int shift_amount)
	{
		sbyte csharp_value = (sbyte)(values [index] << shift_amount);
		sbyte vec_value = vec.GetElement (index);
		if (csharp_value != vec_value) {
			Console.WriteLine ("Error shifting {0} by {1}, ({2} != {3})", values [index], shift_amount, vec_value, csharp_value);
			return false;
		} else {
			return true;
		}
	}

	public static void TestShift ()
	{
		int num_elements = Vector128<sbyte>.Count;
		sbyte[] values = new sbyte[num_elements];

		for (int shift_amount = -9; shift_amount < 9; shift_amount++) {
			Console.WriteLine ("Testing shift {0}", shift_amount);
			bool pass = true;
			for (sbyte i = sbyte.MinValue; i < sbyte.MaxValue && pass; i++) {
				for (int k = 0; k < num_elements; k++)
					values [k] = i;

				Vector128<sbyte> vector = Vector128.Create (values);
				vector = Vector128.ShiftLeft (vector, shift_amount);

				for (int k = 0; k < num_elements && pass; k++)
					pass &= CheckValue (vector, values, k, shift_amount);
			}
			if (pass)
				Console.WriteLine ("OK");
			else
				Console.WriteLine ("NOK");
		}
	}

	public static void Main (string[] args)
	{
		TestShift ();
	}
}
```

tannergooding replied:
>  From what I understand from you, the shifting in c# might by different from the shifting done by the vectorized assembly operations behind the scenes.

We have two different scenarios to consider.

First we have the platform specific hardware intrinsics such as `System.Runtime.Intrinsics.X86.Sse2.ShiftLeftLogical` or `System.Runtime.Intrinsics.Arm.AdvSimd.ShiftLeftLogical`. These intrinsics emit exactly 1 instruction and may do different operations per platform.

We then also have cross platform hardware intrinsics such as `System.Runtime.Intrinsics.Vector128.ShiftLeft` and `System.Runtime.Intrinsics.Vector128<T>.operator <<`. These intrinsics ensure consistent and deterministic behavior across all platforms and ensure that edge cases are handled where applicable.

For the cross platform hardware intrinsics, you can see the software fallback implementation of `operator <<` [here](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Vector64_1.cs#L294-L305). You'll see that it calls [Scalar<T>.ShiftLeft](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Scalar.cs#L796-L854) which in turn does the shifting in C# and therefore ensures that all the shifting is "well-defined" (it masks `shiftCount` against `7`, `15`, or relies on the implicit masking C# emits for `31` and `63`).

We then also have logic for the intrinsic paths to ensure the IR is imported correctly for non-constants and that it gives us the same ultimate behavior of masking the `shiftCount`: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/gentree.cpp#L19552-L19624

> So I wrote a testcase to verify shifting for every single possible sbyte value.

I explained this bit above, but it might have gotten lost. C# (and .NET IL in general) doesn't have "proper" support for small-types (`byte`, `sbyte`, `short`, `ushort`, `bool`, or `char`). They are all implemented in terms of an implicit upcast to `int` and so for `sbyte`, your `(sbyte)(values [index] << shift_amount)` is being emitted in IL as something more like: `(sbyte)((int)(values[index]) << (shift_amount & 31))`

The intrinsics all operate directly with small types and thus have "proper support" for them. The same applies to `generic math` where `byte op byte` produces `byte`, not `int`. Thus, we ensure there remains well-defined logic in the face of small types for what we expose.

BrzVlad replied:
I see. Thanks for the detailed explanation. The behavior was indeed different with or without this PR for those shift amounts. I added the masking within those intrinsics and the outputs seem to be correct now.

BrzVlad committed:
[mono][interp] Remove op_Division

It is actually not used in bcl, it is not really vectorized on any platforms and the codegen for the interp implementation is massive and inefficient.

BrzVlad committed:
[mono][interp] Don't emit intrinsics for unsupported vector types

BrzVlad committed:
[mono][interp] Vector extensions used in these intrinsics are a GNUC extension

kotlarmilos commented on src/mono/mono/mini/interp/interp-simd.c:
```diff
@@ -0,0 +1,568 @@
+
+#define SIZEOF_V128	16
+
+typedef gint64 v128_i8 __attribute__ ((vector_size (16)));
```
Can `SIZEOF_V128` be used instead of const?

kotlarmilos commented on src/mono/mono/mini/interp/transform-simd.c:
```diff
@@ -0,0 +1,465 @@
+				td->ip += 5;
+				return TRUE;
+			} else {
+				// asm ("int $3");
```
Is it intended to be empty?

kotlarmilos commented on src/mono/mono/mini/interp/transform.h:
```diff
@@ -282,6 +282,78 @@ typedef struct
+#else
+#define WRITE32(ip, v) \
+        do { \
+                * (guint32*)(ip) = * (guint32 *)(v); \
```
space before * is missing

BrzVlad commented:
/azp run runtime-extra-platforms

kg commented:
Don't have a full grip on this PR yet to the point that I could checkmark it, but it seems promising so far. I have a basic jiterpreter implementation for it such that there are no trace aborts due to SIMD opcodes.

lambdageek commented:
This is a lot of boilerplate that is very similar to what we do for the JIT/AOT.  Is there some way we can share more between the execution engines?  There's a ton of these SIMD opcodes it would be better if we could generate the code for each execution engine from a single abstract template.  Maybe a python script or something...

/cc @jandupej

tannergooding commented:
> Is there some way we can share more between the execution engines?

RyuJIT handles this via a single header file per architecture: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/hwintrinsiclistxarch.h

We have an entry per intrinsic that tracks most of the table drivable information and a few flags to help cover anything that needs special handling or lookups (which happens through well-defined helper methods/etc).

I imagine a very similar setup would work for `JitIntrinsics -> LlvmIntrinsics` or for `JitIntrinsics -> MonoInstructions`

Mono being C only somewhat restricts the ability for some of this to be shared between RyuJIT and Mono, but that might be fine overall.

jandupej commented:
> This is a lot of boilerplate that is very similar to what we do for the JIT/AOT. Is there some way we can share more between the execution engines? There's a ton of these SIMD opcodes it would be better if we could generate the code for each execution engine from a single abstract template. Maybe a python script or something...
> 
> /cc @jandupej

I agree that would be beneficial. It will take some tuning to get this tabular approach right and applicable to mini, LLVM and interpreter. So my plan is to start small and do it with a single category of operations first (e.g. comparisons). When the final form is ironed out, we will convert existing code to the new form.

jandupej commented on src/mono/mono/mini/interp/interp-simd.c:
```diff
@@ -0,0 +1,566 @@
+
+#ifdef INTERP_ENABLE_SIMD
+
+typedef gint64 v128_i8 __attribute__ ((vector_size (SIZEOF_V128)));
```
I'm not sure this would compile in MSVC. But that's probably not a big issue.

BrzVlad replied:
`INTERP_ENABLE_SIMD` is disabled on msvc due to this reason

kg commented:
Shelving the jiterpreter bits for now, but they're at https://github.com/kg/runtime/commit/a38fa176fe2d4bccbbd827ee4c11005292690c0e.

BrzVlad commented:
There are no issues that I'm aware of with this PR. It needs a final approval so we can get it in and get more perf numbers.

lambdageek commented on src/mono/mono/mini/interp/interp-internals.h:
```diff
@@ -101,6 +102,10 @@ typedef enum {
 
 #define PROFILE_INTERP 0
 
+#if !HOST_BROWSER && __GNUC__
```
```suggestion
#if !HOST_BROWSER && !HOST_WASI && __GNUC__
```

BrzVlad replied:
The real reason for disabling on browser is not because this PR has some problems with wasm, but because this PR leads to jiterpreter trace aborts since support is not yet added. On WASI there is no jiterpreter support so this PR should still have perf gains.

lambdageek commented on src/mono/mono/mini/interp/interp-simd.c:
```diff
@@ -0,0 +1,566 @@
+
+// Shuffle
+
+#define V128_SHUFFLE(eltype, itype) do {					\
```
Curious why you have to do this by hand and not with `__builtin_shuffle` ?

BrzVlad replied:
I simply didn't want to make use of any additional builtins for now, to keep it as simple as possible and reduce potential compatibility issues. Will start adding various builtins in future PR's, probably even arch specific ones. I think `__builtin_shufflevector` has better support and is very powerful, so might be a better candidate.

lambdageek reviewed (approved):
1. I think we should disable WASI
2. Why can't we use `__builtin_shuffle(vec1, mask)`
